Implemented CA1704: Identifiers should be spelled correctly

pakdev opened the pull request:
fixes https://github.com/dotnet/roslyn-analyzers/issues/443 
CA1704 checks that assembly, delegate parameters, members, member parameters, method type parameters, namespaces, types, and type parameters are spelled correctly and “meaningful”, where meaningful means their identifier is more than a single character. To support this analyzer, multiple new non-portable projects had to be created; there's no portable .NET spell checker that I could find. The spell checker itself is ported from the open source StyleCop project because 1. it includes a dictionary and 2. that dictionary (should) match what was used in FxCop (easing the transition to Roslyn).

Like the previous FxCop rule, this analyzer also supports custom dictionaries. They are passed as additional files, and existing codebases can easily transition to this mechanism by adding `<PropertyGroup><AdditionalFileItemNames>$(AdditionalFileItemNames);CodeAnalysisDictionary</AdditionalFileItemNames></PropertyGroup>` to the project file ([more information](https://github.com/dotnet/roslyn/blob/master/docs/analyzers/Using%20Additional%20Files.md)). If the file  path exists and does not update, a caching mechanism will ensure it's only parsed once per Visual Studio instance.

pakdev commented:
I'll get back to this PR pretty soon. TODO:

- [x] Do not report lambda parameters or types deriving from System.EventArgs as unmeaningful
- [x] Remove duplicate classes from the Spelling library (WordParser.cs, WordParserOptions.cs)
- [ ] Do .NET4 projects have to be created now that this project uses .NETStandard?

mavasani commented on src/Spelling/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,207 @@
+        /// <summary>
+        /// Latest date and time of the custom dictionary's last write time/creation time.
+        /// </summary>
+        public DateTime LastUpdated { get; private set; }
```
You don't need such a custom caching infrastructure for storing custom data per-additional file. [AnalysisContext.TryGetValue](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs,211) already provides this functionality. You just create your own [SourceTextValueProvider](http://source.roslyn.io/#Microsoft.CodeAnalysis/DiagnosticAnalyzer/SourceTextValueProvider.cs,820d323c07c3f549) that takes the delegate to compute data for a given additional text, and a value comparer to compare additional text ([default implementation](http://source.roslyn.io/#Microsoft.CodeAnalysis/Text/SourceTextComparer.cs,b6556886ea9ba2b6,references) performs checksum comparison)

pakdev replied:
First, thanks for taking the time to review this. Second, good to know. When I get a chance I'll look into this approach.

pakdev replied:
No longer applies

mavasani commented on src/Spelling/WordParser.cs:
```diff
@@ -0,0 +1,566 @@
+    /// <summary>
+    /// The word parser.
+    /// </summary>
+    public class WordParser
```
We already have a WordParser in our repo: https://github.com/dotnet/roslyn-analyzers/blob/master/src/Utilities/WordParser.cs. How is the new one different from that?

pakdev replied:
No longer applies

mavasani commented on src/Text.Analyzers/Core (Net4)/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -0,0 +1,416 @@
+
+                // Find IdentifierTokens with the syntax tree because SyntaxNodeAction only filters on "methods and above"
+                // Reference: http://stackoverflow.com/a/28004943
+                compilationStartContext.RegisterSyntaxTreeAction(syntaxTreeContext =>
```
This is wrong. `RegisterSyntaxTreeAction` is designed only for reporting syntax diagnostics, where semantic model is not required. If you want to use the semantic model, you should register a semantic model action.

pakdev replied:
What's the difference between my approach and `RegisterSemanticModelAction`'s `context.SemanticModel.SyntaxTree.GetRoot().DescendantToken()`? While I want to use the semantic model, I also want to examine syntax tokens.

mavasani replied:
Syntax tree actions are assumed to be lightweight which do not lead to any binding (expensive computation). This is important especially in the IDE live analysis where the IDE analyzer driver front loads all the syntax tree actions (including compiler's syntax/parse diagnostics).

pakdev replied:
Ahh, thanks for the clarification. Now I need to change another internal analyzer  ;)

pakdev replied:
Switched to Symbol/Operation actions

mavasani commented on src/Text.Analyzers/Core (Net4)/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -0,0 +1,416 @@
+                });
+
+                // Find IdentifierTokens with the syntax tree because SyntaxNodeAction only filters on "methods and above"
+                // Reference: http://stackoverflow.com/a/28004943
```
The referenced issue does not find **symbols** for locals and parameters. However, `SyntaxNodeAction`  will be called back for every single syntax node with matching kind in the tree.

I would propose registering a symbol action for analyzing all top level symbols (including parameters) and then an operation action with `OperationKind.VariableDeclarator` to analyze locals.

pakdev replied:
So `RegisterSymbolAction` does "not find symbols for locals and parameters", but you propose registering a symbol action for "analyzing all top level symbols (**including parameters**)"?

mavasani replied:
You can invoke [this](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Unfactored/Roslyn/Core/ISymbolExtensions.cs#L11) extension methods to get all parameters for member symbols.

pakdev replied:
Cool thanks

pakdev replied:
Hmm, what should I do to get type parameters e.g. `TFoo` in `Method<TFoo>(...)`? Apparently `RegisterSymbolAction` doesn't fire for `SymbolKind.TypeParameter` even though they have symbols...

pakdev replied:
Nevermind; used `IMethodSymbol/INamedTypeSymbol.TypeParameters`

mavasani commented:
Adding @jinujoseph @sharwell for reviews.

Overall, this PR looks very promising to start implementing our naming rules. Thanks a ton @pakdev.

pakdev commented:
Glad I could help, @mavasani. It may take me a month or so to make these changes as I've mostly switched roles, though. If you guys want to cannibalize my work before that, just let me know so I'm not duplicating effort.

mavasani commented:
@pakdev Thanks. I will let @jinujoseph @kuhlenh pitch in for the timelines. If we indeed want to get this rule and other naming rules checked in earlier, we may take over your PR and update it ourselves to fast track it. If so, we will surely let you know to avoid duplicate work.

jinujoseph commented:
@pakdev thanks for the PR. As Manish said we have to plan which shipping release this to target and if early we can take it up and finish the work from our end.

sharwell commented on src/Spelling/SpellCheckerService.cs:
```diff
@@ -0,0 +1,205 @@
+        {
+            if (disposing)
+            {
+                try
```
:bulb:

```csharp
var spellChecker = Interlocked.Exchange(ref _spellChecker, null);
spellChecker?.Dispose();
```

pakdev replied:
No longer applies

sharwell commented on src/Spelling/WindowsSpellChecker.cs:
```diff
@@ -0,0 +1,752 @@
+            StartsSentence = 2
+        }
+
+#pragma warning disable CA1717 // Only FlagsAttribute enums should have plural names
```
💭 Seems like this is a bug in CA1717 if we're still seeing it reported

pakdev replied:
No longer applies

sharwell commented on src/Spelling/WindowsSpellChecker.cs:
```diff
@@ -0,0 +1,752 @@
+                            _id = IntPtr.Zero;
+                        }
+
+                        if (_libraryHandle != IntPtr.Zero)
```
❕ Use safe handles

pakdev replied:
No longer applies

sharwell commented on src/Spelling/WindowsSpellChecker.cs:
```diff
@@ -0,0 +1,752 @@
+                            _lexicons = null;
+                        }
+
+                        if (_id != IntPtr.Zero)
```
❕ Use safe handles

pakdev replied:
No longer applies

sharwell commented on src/Spelling/WindowsSpellChecker.cs:
```diff
@@ -0,0 +1,752 @@
+                    {
+                        if (_lexicons != null)
+                        {
+                            foreach (IntPtr ptr in _lexicons)
```
❕ Use safe handles

pakdev replied:
No longer applies

sharwell commented on src/Spelling/WindowsSpellChecker.cs:
```diff
@@ -0,0 +1,752 @@
+
+                // Needed to clean-up unmanaged resources
+#pragma warning disable CA1821 // Remove empty Finalizers
+                ~Speller()
```
:memo: Remove finalizer after moving to safe handles

pakdev replied:
No longer applies

sharwell commented on src/Spelling/WindowsSpellChecker.cs:
```diff
@@ -0,0 +1,752 @@
+                }
+
+                // Needed to clean-up unmanaged resources
+#pragma warning disable CA1821 // Remove empty Finalizers
```
❗️ This finalizer isn't empty. No suppression is needed.

pakdev replied:
No longer applies

sharwell commented on src/Spelling/WordParser.cs:
```diff
@@ -0,0 +1,566 @@
+
+        private static bool IsLetterWithoutCase(char c)
+        {
+            return (char.IsLetter(c) && !char.IsUpper(c)) && !char.IsLower(c);
```
:bulb: Remove associative parentheses (they hurt readability here)

pakdev replied:
No longer applies

sharwell commented on src/Spelling/WordParser.cs:
```diff
@@ -0,0 +1,566 @@
+            while (IsDigit(Peek()));
+        }
+
+        private void ParseLatex()
```
❔ Why? LaTeX isn't supported in C# code comments.

pakdev replied:
No longer applies

sharwell commented on src/Spelling/Spelling.csproj:
```diff
@@ -0,0 +1,71 @@
+    <Compile Include="WordSpelling.cs" />
+  </ItemGroup>
+  <ItemGroup>
+    <None Include="mssp7en_x86.dll">
```
❗️ This pull request will need to be rewritten so these binaries are not added to source control. A rebase will be required to ensure the files are fully purged.

pakdev replied:
Rewritten

sharwell commented on src/Analyzers.sln:
```diff
@@ -225,6 +225,14 @@ Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "xUnit", "Dependencies\xUnit
 EndProject
 Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "System.Security.Cryptography.Hashing.Algorithms.Analyzers.Setup", "System.Security.Cryptography.Hashing.Algorithms.Analyzers\Setup\System.Security.Cryptography.Hashing.Algorithms.Analyzers.Setup.csproj", "{989CDCD1-9031-4922-895B-8EB4D18D7A2D}"
 EndProject
+Project("{FAE04EC0-301F-11D3-BF4B-00C04F79EFBC}") = "Text.Analyzers.Core (Net4)", "Text.Analyzers\Core (Net4)\Text.Analyzers.Core (Net4).csproj", "{703B2DE5-8131-4904-9CDD-BBD7708D9D36}"
```
❗️ Remove spaces and special characters from the project paths

pakdev replied:
No longer applies

sharwell commented on src/Text.Analyzers/Core (Net4)/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -0,0 +1,416 @@
+        private NamingService _namingService;
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(
+            AssemblyRule,
```
:memo: Only one diagnostic is required here since they all have the same ID

pakdev replied:
No longer applies

sharwell commented on src/Text.Analyzers/Core (Net4)/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -0,0 +1,416 @@
+                _namingService.LoadDictionaries(compilationStartContext.Options.AdditionalFiles, compilationStartContext.CancellationToken);
+
+                // CompilationStartAction doesn't support reporting diagnostics, so we have to check the assembly name in CompilationEndAction
+                compilationStartContext.RegisterCompilationEndAction(compilationEndContext =>
```
:memo: `RegisterCompilationEndAction` doesn't do anything (dotnet/roslyn#11750), so this should be rewritten with an alternate approach that provides the desired behavior consistently.

pakdev replied:
Dropped

sharwell commented on src/Text.Analyzers/Core (Net4)/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -0,0 +1,416 @@
+
+        public override void Initialize(AnalysisContext analysisContext)
+        {
+            analysisContext.EnableConcurrentExecution();
```
❗️ Instance state variables must be removed

pakdev replied:
No longer applies

sharwell commented on src/Text.Analyzers/Core (Net4)/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -0,0 +1,416 @@
+
+        private Diagnostic GetUnmeaningfulIdentifierDiagnostic(SyntaxToken token)
+        {
+            return GetDiagnostic(
```
:memo: The delegates here need to be rewritten to avoid allocating every time this method is called

pakdev replied:
No longer applies

sharwell commented on src/Spelling/WordParser.cs:
```diff
@@ -0,0 +1,566 @@
+        /// <returns>
+        /// The System.String.
+        /// </returns>
+        public string NextWord()
```
❕ Needs to be rewritten to return a non-allocating view of the original string, or it will not scale

pakdev replied:
No longer applies

sharwell reviewed (dismissed):
This pull request is going to take a substantial amount of work to update to meet the current coding style/requirements of this repository. However, I would recommend holding off on these changes for now. I have serious concerns about re-introducing this feature under the original rule ID (CA1704), so I'm going to request an internal review to see where we are on overall goals.

In the event this does not get included *here*, it could still be developed as a separate feature which is compatible (and can be used at the same time as) other rules defined in this repository. I know several users working with other third-party analyzers have requested similar functionality as well.

ltsstar commented:
I just noticed this PR going on for a long time. We would be glad to have the Text.Analyzer tool running to check our code recurrently for spelling errors!
So please keep working on that PR, that the code gets into the master at one point!

mavasani commented:
Tagging @jinujoseph @vatsalyaagrawal - we have had multiple requests now requesting porting the naming rules, some of them also marking as blockers for FxCop migration. We should budget implementing this soon, starting with getting this PR moving again.

pakdev commented:
@mavasani, I'm finally revisiting this PR. So far, I've reworked it to use `RegisterSymbolAction`. Before I address @sharwell's feedback about using SafeHandles, though, I want to confirm we still want to use the Office dictionary. What do you guys think of [WeCantSpell.Hunspell](https://github.com/aarondandy/WeCantSpell.Hunspell/)? Hunspell "is the spell checker of LibreOffice, OpenOffice.org, Mozilla Firefox 3 & Thunderbird, Google Chrome."

genlu commented:
> What do you guys think of WeCantSpell.Hunspell? Hunspell "is the spell checker of LibreOffice, OpenOffice.org, Mozilla Firefox 3 & Thunderbird, Google Chrome..."

We should definitely evaluate this, looks promising :)

pakdev commented:
It turns out Hunspell offers an "unmunch" command that generates a list of all words. When this list is added to a hash set, spell checking simply becomes `_wordList.Contains(word)` - no dependencies required. Profiling reveals that it's slightly slower than Microsoft's unmanaged dictionary, but the simplicity more than makes up for it. I'll push an implementation when I get some time.

mavasani commented:
We are getting bunch of requests for the spell checker rules. @pakdev Any change you have time for this in near future? Otherwise, I might give that package a try.

pakdev commented:
@mavasani I wrote one internally for my company. I'll copy and paste it into this project, quickly fix what I can, and update this PR.

pakdev commented:
PR updated. Hopefully you can use some of my implementation, though I had to comment out most of the tests. It seems test verification has greatly diverged since the last time I was in your codebase. I also implemented my own `TestFile`/`AutoTestFile` scheme to bundle project names with source files and allow "markers" (e.g. <?>) to denote locations of errors vs. explicit row/column numbers. It probably wouldn't be too hard to adapt them to the current conventions, but it doesn't seem like [adding additional files to additional projects](https://github.com/dotnet/roslyn-sdk/blob/master/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/SolutionState.cs#L51) is supported.

sharwell commented on src/Test.Utilities/TestAdditionalDocument.cs:
```diff
@@ -0,0 +1,41 @@
+    /// <summary>
+    /// An implementation of AdditionalText meant specifically for unit-testing.
+    /// </summary>
+    public class TestAdditionalDocument : AdditionalText
```
📝 This is not needed with the new test library

pakdev replied:
Do you have an example/documentation?

sharwell replied:
https://github.com/dotnet/roslyn-analyzers/blob/dba3183193df36d666fc48c0eac0d2d640d0041d/src/PublicApiAnalyzers/UnitTests/DeclarePublicApiAnalyzerTests.cs#L105-L113

mavasani replied:
+1

sharwell replied:
@mavasani I'm not sure who is currently working on this, but I was expecting this to get removed

pakdev replied:
Removed

sharwell commented on src/Utilities/Compiler/Extensions/SourceTextExtensions.cs:
```diff
@@ -0,0 +1,47 @@
+        /// <param name="text">Abstraction for an additional file's contents.</param>
+        /// <param name="parser">Function that will parse <paramref name="text"/> into <typeparamref name="T"/>.</param>
+        /// <returns>Output from <paramref name="parser"/>.</returns>
+        public static T Parse<T>(this SourceText text, Func<StreamReader, T> parser)
```
❕  Use `AnalysisContext.TryGetValue` instead. This method can be removed.

pakdev replied:
Wouldn't I still need code to read/parse SourceText?

sharwell replied:
That's what [`TryGetValue`](https://github.com/dotnet/roslyn/blob/1309e6b088a884104057b7614389493d436b3aee/src/Compilers/Core/Portable/DiagnosticAnalyzer/DiagnosticAnalysisContext.cs#L226-L236) does.

pakdev replied:
Haha, `TryGetValue` calls `SourceText.[Lines|Write]`? In my code, `TryGetValue` takes `SourceText` as the "key" and either `_xmlDictionaryProvider` or `_dicDictionaryProvider` as the `SourceTextValueProvider`. These providers read/parse the underlying text from `SourceText` with this method and map it to a `CodeAnalysisDictionary`. It's this mapped value that `TryGetValue` returns either by cache or calculation.

mavasani replied:
@sharwell is correct. If multiple analyzers end up requiring converting the same source text instance to dictionary value, using `AnalysisContext.TryGetValue` will ensure the compute is shared.

mavasani replied:
Ah, so reading the code, this seems like the core helper method used by the `AnalysisContext.TryGetValue` callback, correct? If so, this seems reasonable to me.

pakdev replied:
Right

Evangelink commented:
@pakdev Hey! Thanks for the good work, do you think you would have some time to move forward on this PR?

pakdev commented:
I don't have a ton of spare time, but I'll do what I can if someone from MS (@sharwell?) confirms this approach can be merged once the remaining issues are addressed.

mavasani commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            s_localizableTitle,
+            s_localizableMessageFileParse,
+            DiagnosticCategory.Naming,
+            DiagnosticSeverity.Error,
```
No analyzer rule should be error by default - please switch this to warning.

pakdev replied:
Well that's annoying that the comment becomes orphaned on rebase...anyway, switched to warning

mavasani commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            }
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext compilationStartContext)
```
Please move this method right below `Initialize`

pakdev replied:
Done

mavasani commented on src/Utilities/Compiler/AdditionalFileProvider.cs:
```diff
@@ -0,0 +1,55 @@
+    {
+        private readonly ImmutableArray<AdditionalText> _additionalFiles;
+
+        internal AdditionalFileProvider(params AdditionalText[] additionalFiles)
```
This constructor seems unused.

pakdev replied:
It was used by AdditionalFileProviderTests that I originally forgot to include. I've now included them and just used the ImmutableArray override

mavasani commented on src/Text.Analyzers/UnitTests/IdentifiersShouldBeSpelledCorrectlyTests.cs:
```diff
@@ -1,8 +1,473 @@
+                    .WithArguments("Program", "Program"));
+        }
+
+        // [Fact]
```
Tests need to be uncommented or removed.

mavasani replied:
Seems like all diagnostic verification tests are commented out - this will certainly block this PR.

pakdev replied:
At least one of my tests needs to modify the assembly name so it's misspelled/unmeaningful, but it doesn't appear this is possible in [AnalyzerTest`1.cs](https://github.com/dotnet/roslyn-sdk/blob/1f045832e3db8653f9c3b4baecb2d6d133e92096/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/AnalyzerTest%601.cs#L1100). Is that correct?

I also need `AdditionalFiles` to be associated with specific projects, but they're just listed in [SolutionState.cs](https://github.com/dotnet/roslyn-sdk/blob/master/src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/SolutionState.cs) vs. ProjectState.cs.

pakdev replied:
Still working on converting the tests...

pakdev replied:
Tests are "done" (see skipped)

mavasani commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+        {
+            var assemblyType = typeof(IdentifiersShouldBeSpelledCorrectlyAnalyzer);
+            var assembly = assemblyType.GetTypeInfo().Assembly;
+            var dictionary = $"{assemblyType.Namespace}.Dictionary.dic";
```
Please add documentation on how this dictionary.dic file was generated, and how it can be refreshed, if required in future.

pakdev replied:
Done

mavasani commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                CodeAnalysisDictionary CreateDictionaryFromAdditionalText(AdditionalText additionalFile)
+                {
+                    var text = additionalFile.GetText(compilationStartContext.CancellationToken);
+                    var isXml = additionalFile.Path.EndsWith("xml", StringComparison.OrdinalIgnoreCase);
```
Can we instead check the file extension?

sharwell replied:
I would still prefer the `.`. Otherwise this reads like a code smell.

```suggestion
                    var isXml = additionalFile.Path.EndsWith(".xml", StringComparison.OrdinalIgnoreCase);
```

pakdev replied:
Added .

mavasani commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -1,24 +1,34 @@
     /// </summary>
-    public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAnalyzer
+    [DiagnosticAnalyzer(LanguageNames.CSharp, LanguageNames.VisualBasic)]
+    public class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAnalyzer
```
```suggestion
    public sealed class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAnalyzer
```

pakdev replied:
Can I remove the CSharpIdentifiersShouldBeSpelledCorrectly/BasicIdentifiersShouldBeSpelledCorrectly classes then?

Evangelink replied:
Either you need C#/VB.NET specific APIs in which case this class should be `abstract` or you can do everything from the language agnostic layer in which case this should be `sealed` and you don't need the sub-classes (you can remove them).

pakdev replied:
Done

mavasani commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,134 @@
+    /// </Remarks>
+    internal class CodeAnalysisDictionary
+    {
+        protected CodeAnalysisDictionary(XDocument document)
```
This constructor feels weird. I think it would be better if all the processing/loading happened in the static factory methods on `CodeAnalysisDictionary` and the constructors only accept the final immutable data. This is the pattern used everywhere across this repo.

mavasani replied:
Also why are these constructors all protected?

pakdev replied:
Done and removed protected

mavasani commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,134 @@
+
+        protected CodeAnalysisDictionary(IEnumerable<string> recognizedWords)
+        {
+            RecognizedWords.UnionWith(recognizedWords);
```
Same here - lets change all the constructors to take the final immutable dictionaries as input parameters.

pakdev replied:
Done

mavasani commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,134 @@
+
+        private void LoadWordsFromDocument(XDocument document)
+        {
+            RecognizedWords.UnionWith(GetWordsOrExceptions("Recognized", "Word"));
```
Probably add a TODO comment here to handle words marked as Deprecated and handling of Acronyms as per the documentation here: https://docs.microsoft.com/visualstudio/code-quality/ca1704?view=vs-2019#to-add-words-to-a-custom-dictionary

pakdev replied:
Done

mavasani commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                    break;
+
+                case SymbolKind.NamedType:
+                    foreach (var location in symbol.Locations)
```
This will end up reporting duplicate diagnostics per each partial declaration. Instead we just report diagnostic for symbol's first location. You should be able to change the method to return `Diagnostic` instead of `IEnumerable<Diagnostic>`

pakdev replied:
Report first location now

mavasani commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                    break;
+
+                case SymbolKind.NamedType:
+                    foreach (var location in symbol.Locations)
```
Same here, please report only on symbol's first location.

pakdev replied:
Done

mavasani commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            }
+        }
+
+        private static IEnumerable<Diagnostic> GetMisspelledWordDiagnostics(ISymbol symbol, string misspelledWord)
```
```suggestion
        private static Diagnostic GetMisspelledWordDiagnostic(ISymbol symbol, string misspelledWord)
```

mavasani replied:
Actually the method can be changed to `private static void ReportMisspelledWordDiagnostic(ISymbol symbol, string misspelledWord, Action<Diagnostic> addDiagnostic)`

pakdev replied:
Kept it "Get" but reported it sooner

mavasani commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+        private static string RemovePrefixIfPresent(string prefix, string name)
+            => name.StartsWith(prefix, StringComparison.Ordinal) ? name.Substring(1) : name;
+
+        private static IEnumerable<Diagnostic> GetUnmeaningfulIdentifierDiagnostics(ISymbol symbol, string symbolName)
```
```suggestion
        private static Diagnostic GetUnmeaningfulIdentifierDiagnostic(ISymbol symbol, string symbolName)
```

pakdev replied:
Done

mavasani reviewed (changes requested):
Overall, the PR seems in good place. Needs tests and few cleanups.

sharwell commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,134 @@
+        /// </Recognized>
+        /// </code>
+        /// </example>
+        public HashSet<string> RecognizedWords { get; protected set; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```
```suggestion
        public HashSet<string> RecognizedWords { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```

pakdev replied:
Done

sharwell commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,134 @@
+        /// </Unrecognized>
+        /// </code>
+        /// </example>
+        public HashSet<string> UnrecognizedWords { get; protected set; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```
```suggestion
        public HashSet<string> UnrecognizedWords { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```

pakdev replied:
Done

sharwell commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,134 @@
+        /// <returns>A new instance of this class with recognized words loaded.</returns>
+        public static CodeAnalysisDictionary CreateFromDic(StreamReader streamReader)
+        {
+            var words = new Collection<string>();
```
```suggestion
            var words = new List<string>();
```

pakdev replied:
Done

sharwell commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,134 @@
+        /// Returns a copy of the current instance.
+        /// </summary>
+        /// <returns></returns>
+        public CodeAnalysisDictionary Clone() => new CodeAnalysisDictionary(this);
```
💭 If we make `CodeAnalysisDictionary` immutable, this is no longer required

pakdev replied:
Done

sharwell commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,134 @@
+            return this;
+        }
+
+        private static IEnumerable<T> GetValues<T>(IEnumerable<XElement> elements, Func<XElement, T> extractor)
+            => elements.Select(extractor);
```
💡 Inlining this would improve readability

pakdev replied:
Agreed - done

sharwell commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,134 @@
+            RecognizedWords.UnionWith(GetWordsOrExceptions("Recognized", "Word"));
+            UnrecognizedWords.UnionWith(GetWordsOrExceptions("Unrecognized", "Word"));
+
+            IEnumerable<string> GetWordsOrExceptions(string section, string property) =>
+                GetValues(GetElements(document, section, property), ExtractInnerText);
```
```suggestion
            IEnumerable<string> GetWordsOrExceptions(string section, string property)
                => GetValues(GetElements(document, section, property), ExtractInnerText);
```

pakdev replied:
Inlined

sharwell commented on src/Text.Analyzers/Core/TextAnalyzersResources.resx:
```diff
@@ -174,4 +174,10 @@
+    <value>Error parsing dictionary '{0}': {1}</value>
+  </data>
+  <data name="IdentifiersShouldBeSpelledCorrectlyMessageVariable" xml:space="preserve">
+    <value>Correct the spelling of '{0}' in variable name {1}.</value>
```
```suggestion
    <value>Correct the spelling of '{0}' in variable name '{1}'.</value>
```

pakdev replied:
Added quotes around them all

sharwell commented on src/Text.Analyzers/UnitTests/IdentifiersShouldBeSpelledCorrectlyTests.cs:
```diff
@@ -1,8 +1,473 @@
+        public static IEnumerable<object[]> MisspelledMembers
+            => new[]
+            {
+                new object[] { CreateTypeWithConstructor("<?>Clazz", isStatic: false), "Clazz", "Clazz.Clazz()" },
```
😕 I don't understand the use of `<?>`. Maybe ported Java code?

pakdev replied:
It was my version of markupKey

sharwell commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            s_localizableMessageFileParse,
+            DiagnosticCategory.Naming,
+            DiagnosticSeverity.Error,
+            isEnabledByDefault: false);
```
💭 It seems like this should be enabled by default

pakdev replied:
Done

sharwell commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            var assembly = assemblyType.GetTypeInfo().Assembly;
+            var dictionary = $"{assemblyType.Namespace}.Dictionary.dic";
+
+            using var stream = assembly.GetManifestResourceStream(dictionary);
```
💡 The embedded resource can be exposed through the .resx file without needing to manually include the resource name.

pakdev replied:
TIL - that's much nicer; done

sharwell commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            }
+
+            compilationStartContext.RegisterOperationAction(AnalyzeVariable, OperationKind.VariableDeclarator);
+            compilationStartContext.RegisterCompilationEndAction(AnalyzeAssembly);
```
❕ This analyzer should not include a compilation end action. Assembly diagnostics need to be reported elsewhere.

mavasani replied:
There is no other alternative - assembly level diagnostics have to go into compilation level actions. We already have many analyzers doing such. They will be build-only diagnostics, but that is fine.

pakdev replied:
Kept it

sharwell commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -34,162 +44,541 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                void ReportFileParseDiagnostic(string filePath, string message)
+                {
+                    var diagnostic = Diagnostic.Create(FileParseRule, Location.None, filePath, message);
+                    compilationStartContext.RegisterCompilationEndAction(x => x.ReportDiagnostic(diagnostic));
```
❕ No compilation end actions

pakdev replied:
How am I supposed to report diagnostics from CompilationStartActions? Alternatively, which action should I use to do and report on one operation e.g. checking assembly spelling/loading additional files?

mavasani replied:
I think this is fine - assembly level diagnostics will need to be reported at CompilationEndAction or CompilationAction, either one is fine. We have recently added a new API for additional file actions, but that is not yet available, so even those can go in the same compilation level actions.

sharwell replied:
An acceptable alternative would be splitting the analyzer such that the analyzer that requires Compilation End actions is separate. This will allow the normal analyzer to fully function without FSA enabled but still allow the CEA to report assembly diagnostics on build.

mavasani replied:
Such a refactoring is no longer required. The IDE now treats each diagnostic descriptor separately for build-live de-duping, this is not an analyzer level concept anymore. I think a simpler approach would be to create a new diagnostic ID for diagnostics reported at compilation end, and then adding the CompilationEnd custom tag to this descriptor.

sharwell replied:
> each diagnostic descriptor separately

Does this apply for the _descriptor_ or the _ID_?

mavasani replied:
It is at diagnostic ID level, it is not possible to distinguish at descriptor level as diagnostics coming from build only have an associated ID, and we cannot distinguish which action/descriptor maps to the diagnostic.

pakdev replied:
So it sounds like nothing needs to change here, right?

mavasani replied:
Yes, I think you are good here.

mavasani commented:
@pakdev Would you have time to address the review feedback? Please let me know, thanks!

pakdev commented:
@mavasani I'll try my best between now and the end of next week

pakdev commented:
@mavasani Feedback addressed; see my question

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+        private static readonly CodeAnalysisDictionary _mainDictionary = GetMainDictionary();
+
+        internal static DiagnosticDescriptor FileParseRule = new DiagnosticDescriptor(
+            RuleId,
```
You need to add this to the release file (there is a code-fix for this).

mavasani replied:
@pakdev If you switch the descriptor creation to use the helper like below, it will likely fix your build error: https://github.com/dotnet/roslyn-analyzers/blob/edf95f72a20ed1c090a1639c8ec85bb5854fe09e/src/Text.Analyzers/Core/EnumsShouldHavePluralNames.cs#L43-L52

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            s_localizableTitle,
+            s_localizableMessageFileParse,
+            DiagnosticCategory.Naming,
+            DiagnosticSeverity.Warning,
```
@mavasani will correct me but my understanding is that new rules should never be set to `warning` before any use validation (ensure there is a low level of FPs).

mavasani replied:
@Evangelink These spell checker rules will go into a completely separate experimental package (Text.Analyzers), so it should be fine to have these as warnings by default. If we start shipping these with the .NET SDK, they will have to be bumped down.

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: false,
+            description: s_localizableDescription,
+            helpLinkUri: "https://docs.microsoft.com/visualstudio/code-quality/ca1704",
```
I don't think the help link URL should be set manually.

mavasani replied:
I think it is fine for rules in this experimental package. This may change in future.

pakdev replied:
Hmm, already removed. Do you want them back?

mavasani replied:
Yes, please set the help link as it already exists.

pakdev replied:
I'm no longer able to after switching to DiagnosticDescriptorHelper.Create per [this comment[(https://github.com/dotnet/roslyn-analyzers/pull/1111#discussion_r556584821)

mavasani replied:
Yes, the helper you used already passes a help link so you are all good.

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
-        //ImmutableArray.Create(AssemblyRule, NamespaceRule, TypeRule, MemberRule, MemberParameterRule, DelegateParameterRule, TypeTypeParameterRule, MethodTypeParameterRule, AssemblyMoreMeaningfulNameRule, NamespaceMoreMeaningfulNameRule, TypeMoreMeaningfulNameRule, MemberMoreMeaningfulNameRule, MemberParameterMoreMeaningfulNameRule, DelegateParameterMoreMeaningfulNameRule, TypeTypeParameterMore ...
-
-#pragma warning disable RS1025 // Configure generated code analysis
+        private static readonly SourceTextValueProvider<CodeAnalysisDictionary> _xmlDictionaryProvider = new SourceTextValueProvider<CodeAnalysisDictionary>(ParseXmlDictionary);
```
```suggestion
        private static readonly SourceTextValueProvider<CodeAnalysisDictionary> s_xmlDictionaryProvider = new SourceTextValueProvider<CodeAnalysisDictionary>(ParseXmlDictionary);
```

pakdev replied:
Prefixed all static fields with `s_`

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
-
-#pragma warning disable RS1025 // Configure generated code analysis
+        private static readonly SourceTextValueProvider<CodeAnalysisDictionary> _xmlDictionaryProvider = new SourceTextValueProvider<CodeAnalysisDictionary>(ParseXmlDictionary);
+        private static readonly SourceTextValueProvider<CodeAnalysisDictionary> _dicDictionaryProvider = new SourceTextValueProvider<CodeAnalysisDictionary>(ParseDicDictionary);
```
```suggestion
        private static readonly SourceTextValueProvider<CodeAnalysisDictionary> s_dicDictionaryProvider = new SourceTextValueProvider<CodeAnalysisDictionary>(ParseDicDictionary);
```

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
-#pragma warning disable RS1025 // Configure generated code analysis
+        private static readonly SourceTextValueProvider<CodeAnalysisDictionary> _xmlDictionaryProvider = new SourceTextValueProvider<CodeAnalysisDictionary>(ParseXmlDictionary);
+        private static readonly SourceTextValueProvider<CodeAnalysisDictionary> _dicDictionaryProvider = new SourceTextValueProvider<CodeAnalysisDictionary>(ParseDicDictionary);
+        private static readonly CodeAnalysisDictionary _mainDictionary = GetMainDictionary();
```
```suggestion
        private static readonly CodeAnalysisDictionary s_mainDictionary = GetMainDictionary();
```

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            analysisContext.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext compilationStartContext)
```
As far as I can tell from GitHub, this could be marked as `static`.

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            {
+                var fileProvider = AdditionalFileProvider.FromOptions(compilationStartContext.Options);
+                return fileProvider.GetMatchingFiles(@"(?:dictionary|custom).*?\.(?:xml|dic)$")
+                    .Select(CreateDictionaryFromAdditionalText)
```
There is a performance hit for method group, I don't know if it matters a lot for analyzer.

```suggestion
                    .Select(x => CreateDictionaryFromAdditionalText(x))
```

pakdev replied:
This is news to me; do you have some documentation? The answers on [this question](https://stackoverflow.com/questions/6662532/difference-between-lambda-expression-and-method-group) seem to suggest the opposite.

Evangelink replied:
Yeah that's a weird thing. Here is the a link to my [SO question](https://stackoverflow.com/q/53208516/4625433) and I know there was a good blog post (I think from JetBrains) but I can't seem to find it back...

pakdev replied:
That is interesting, but based on your findings, we shouldn't cross the 1M iteration boundary where this might be a concern.

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                var fileProvider = AdditionalFileProvider.FromOptions(compilationStartContext.Options);
+                return fileProvider.GetMatchingFiles(@"(?:dictionary|custom).*?\.(?:xml|dic)$")
+                    .Select(CreateDictionaryFromAdditionalText)
+                    .Where(x => x != null)
+                    .ToList();
```
```suggestion
                    .Where(x => x != null);
```

pakdev replied:
Without `.ToList`,  this will be enumerated twice and cause my FileParseError test to fail (2 diagnostics)

Evangelink replied:
I don't understand why, as far as I can see it's called only once from the compilation start so it should be ok.

pakdev replied:
I set `dictionaries` to the value of `ReadDictionaries()`, which is the enumerable. Then I pass `dictionaries.Concat(s_mainDictionary)` into `CodeAnalysisDictionary.CreateFromDictionaries`, which does the following:

```
var recognizedWords = dictionaries.Select(x => x.RecognizedWords).Aggregate((x, y) => x.Union(y));
var unrecognizedWords = dictionaries.Select(x => x.UnrecognizedWords).Aggregate((x, y) => x.Union(y));
```

As you can see, the value of `dictionaries` is enumerated twice.

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+
+                CodeAnalysisDictionary CreateDictionaryFromAdditionalText(AdditionalText additionalFile)
+                {
+                    var text = additionalFile.GetText(compilationStartContext.CancellationToken);
```
Let's avoid captured variables.

pakdev replied:
Do you say this for performance reasons? `compilationStartContext` will only be captured once per compilation, and passing it as a parameter will raise an `RS1013: 'compilationStartContext' does not register any analyzer actions, except for a 'CompilationEndAction'...`

Evangelink replied:
@mavasani what's your preference here?

mavasani replied:
I am fine either way given this is a local function. Generally we prefer static local functions, but if that makes the code more messy, captures is fine.

pakdev replied:
Keeping it captures so I don't have to deal with RS1013

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+
+            IEnumerable<CodeAnalysisDictionary> ReadDictionaries()
+            {
+                var fileProvider = AdditionalFileProvider.FromOptions(compilationStartContext.Options);
```
Let's avoid captured variables.

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                SymbolKind.Field,
+                SymbolKind.Parameter);
+
+            IEnumerable<CodeAnalysisDictionary> ReadDictionaries()
```
Could be marked as `static`.

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                    .Where(x => x != null)
+                    .ToList();
+
+                CodeAnalysisDictionary CreateDictionaryFromAdditionalText(AdditionalText additionalFile)
```
Could be marked as `static`.

pakdev replied:
Depends if we pass in parameters or capture variables

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                var symbolName = symbol.Name;
+                switch (symbol)
+                {
+                    case IFieldSymbol field:
```
```suggestion
                    case IFieldSymbol:
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            IEnumerable<string> GetMisspelledWords(string symbolName)
+            {
+                var parser = new WordParser(symbolName, WordParserOptions.SplitCompoundWords);
+                if (parser.PeekWord() != null)
+                {
+                    var word = parser.NextWord();
+                    if (word == null)
+                    {
+                        yield break;
+                    }
+
+                    do
+                    {
+                        if (IsWordAcronym(word) || IsWordNumeric(word) || IsWordSpelledCorrectly(word))
+                        {
+                            continue;
+                        }
+
+                        yield return word;
+                    }
+                    while ((word = parser.NextWord()) != null);
+                }
```
```suggestion
                while ((word = parser.NextWord()) != null)
                {
                    if (!IsWordAcronym(word) && !IsWordNumeric(word) && !IsWordSpelledCorrectly(word))
                    {
                        yield return word;
                    }                    
                }
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+        private static CodeAnalysisDictionary ParseDicDictionary(SourceText text)
+            => text.Parse(CodeAnalysisDictionary.CreateFromDic);
+
+        private static string RemovePrefixIfPresent(string prefix, string name)
+            => name.StartsWith(prefix, StringComparison.Ordinal) ? name[1..] : name;
```
All use of this method only go for a `char` prefix

```suggestion
        private static string RemovePrefixIfPresent(char prefix, string name)
            => name.Length > 0 && name[0] == prefix ? name[1..] : name;
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            return symbol.Kind switch
+            {
+                SymbolKind.Assembly => Diagnostic.Create(AssemblyRule, Location.None, misspelledWord, symbol.Name),
+                SymbolKind.Namespace => Diagnostic.Create(NamespaceRule, symbol.Locations.First(), misspelledWord, symbol.ToDisplayString()),
```
```suggestion
                SymbolKind.Namespace => symbol.CreateDiagnostic(NamespaceRule, misspelledWord, symbol.ToDisplayString()),
```

pakdev replied:
Are you suggesting to not include location?

Evangelink replied:
I am just suggesting to use extension method [CreateDiagnostic](https://github.com/dotnet/roslyn-analyzers/blob/master/src/Utilities/Compiler/Extensions/DiagnosticExtensions.cs#L76) we have defined and already take care of reporting with the right location(s).

pakdev replied:
Super cool; love it

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            {
+                SymbolKind.Assembly => Diagnostic.Create(AssemblyRule, Location.None, misspelledWord, symbol.Name),
+                SymbolKind.Namespace => Diagnostic.Create(NamespaceRule, symbol.Locations.First(), misspelledWord, symbol.ToDisplayString()),
+                SymbolKind.NamedType => Diagnostic.Create(TypeRule, symbol.Locations.First(), misspelledWord, symbol.ToDisplayString()),
```
```suggestion
                SymbolKind.NamedType => symbol.CreateDiagnostic(TypeRule, misspelledWord, symbol.ToDisplayString()),
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                SymbolKind.Assembly => Diagnostic.Create(AssemblyRule, Location.None, misspelledWord, symbol.Name),
+                SymbolKind.Namespace => Diagnostic.Create(NamespaceRule, symbol.Locations.First(), misspelledWord, symbol.ToDisplayString()),
+                SymbolKind.NamedType => Diagnostic.Create(TypeRule, symbol.Locations.First(), misspelledWord, symbol.ToDisplayString()),
+                SymbolKind.Method or SymbolKind.Property or SymbolKind.Event or SymbolKind.Field => Diagnostic.Create(MemberRule, symbol.Locations.First(), misspelledWord, symbol.ToDisplayString()),
```
```suggestion
                SymbolKind.Method or SymbolKind.Property or SymbolKind.Event or SymbolKind.Field => symbol.CreateDiagnostic(MemberRule, misspelledWord, symbol.ToDisplayString()),
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                SymbolKind.NamedType => Diagnostic.Create(TypeRule, symbol.Locations.First(), misspelledWord, symbol.ToDisplayString()),
+                SymbolKind.Method or SymbolKind.Property or SymbolKind.Event or SymbolKind.Field => Diagnostic.Create(MemberRule, symbol.Locations.First(), misspelledWord, symbol.ToDisplayString()),
+                SymbolKind.Parameter => symbol.ContainingType.TypeKind == TypeKind.Delegate
+                    ? Diagnostic.Create(DelegateParameterRule, symbol.Locations.First(), symbol.ContainingType.ToDisplayString(), misspelledWord, symbol.Name)
```
```suggestion
                    ? symbol.CreateDiagnostic(DelegateParameterRule, symbol.ContainingType.ToDisplayString(), misspelledWord, symbol.Name)
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                SymbolKind.Method or SymbolKind.Property or SymbolKind.Event or SymbolKind.Field => Diagnostic.Create(MemberRule, symbol.Locations.First(), misspelledWord, symbol.ToDisplayString()),
+                SymbolKind.Parameter => symbol.ContainingType.TypeKind == TypeKind.Delegate
+                    ? Diagnostic.Create(DelegateParameterRule, symbol.Locations.First(), symbol.ContainingType.ToDisplayString(), misspelledWord, symbol.Name)
+                    : Diagnostic.Create(MemberParameterRule, symbol.Locations.First(), symbol.ContainingSymbol.ToDisplayString(), misspelledWord, symbol.Name),
```
```suggestion
                    : symbol.CreateDiagnostic(MemberParameterRule, symbol.ContainingSymbol.ToDisplayString(), misspelledWord, symbol.Name),
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                SymbolKind.TypeParameter => symbol.ContainingSymbol.Kind == SymbolKind.Method
+                    ? Diagnostic.Create(MethodTypeParameterRule, symbol.Locations.First(), symbol.ContainingSymbol.ToDisplayString(), misspelledWord, symbol.Name)
+                    : Diagnostic.Create(TypeTypeParameterRule, symbol.Locations.First(), symbol.ContainingSymbol.ToDisplayString(), misspelledWord, symbol.Name),
+                SymbolKind.Local => Diagnostic.Create(VariableRule, symbol.Locations.First(), misspelledWord, symbol.ToDisplayString()),
```
```suggestion
                SymbolKind.Local => symbol.CreateDiagnostic(VariableRule, misspelledWord, symbol.ToDisplayString()),
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                    ? Diagnostic.Create(DelegateParameterRule, symbol.Locations.First(), symbol.ContainingType.ToDisplayString(), misspelledWord, symbol.Name)
+                    : Diagnostic.Create(MemberParameterRule, symbol.Locations.First(), symbol.ContainingSymbol.ToDisplayString(), misspelledWord, symbol.Name),
+                SymbolKind.TypeParameter => symbol.ContainingSymbol.Kind == SymbolKind.Method
+                    ? Diagnostic.Create(MethodTypeParameterRule, symbol.Locations.First(), symbol.ContainingSymbol.ToDisplayString(), misspelledWord, symbol.Name)
```
```suggestion
                    ? symbol.CreateDiagnostic(MethodTypeParameterRule, symbol.ContainingSymbol.ToDisplayString(), misspelledWord, symbol.Name)
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+                    : Diagnostic.Create(MemberParameterRule, symbol.Locations.First(), symbol.ContainingSymbol.ToDisplayString(), misspelledWord, symbol.Name),
+                SymbolKind.TypeParameter => symbol.ContainingSymbol.Kind == SymbolKind.Method
+                    ? Diagnostic.Create(MethodTypeParameterRule, symbol.Locations.First(), symbol.ContainingSymbol.ToDisplayString(), misspelledWord, symbol.Name)
+                    : Diagnostic.Create(TypeTypeParameterRule, symbol.Locations.First(), symbol.ContainingSymbol.ToDisplayString(), misspelledWord, symbol.Name),
```
```suggestion
                    : symbol.CreateDiagnostic(TypeTypeParameterRule, symbol.ContainingSymbol.ToDisplayString(), misspelledWord, symbol.Name),
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+        private static string RemovePrefixIfPresent(string prefix, string name)
+            => name.StartsWith(prefix, StringComparison.Ordinal) ? name[1..] : name;
+
+        private static Diagnostic GetMisspelledWordDiagnostic(ISymbol symbol, string misspelledWord)
```
You could use expression body here instead.

sharwell replied:
➡️ Marking resolved per https://github.com/dotnet/roslyn-analyzers/pull/1111#discussion_r548020530

Evangelink commented on src/Text.Analyzers/Core/IdentifiersShouldBeSpelledCorrectly.cs:
```diff
@@ -33,162 +43,465 @@ public abstract class IdentifiersShouldBeSpelledCorrectlyAnalyzer : DiagnosticAn
+            };
+        }
+
+        private static Diagnostic GetUnmeaningfulIdentifierDiagnostic(ISymbol symbol, string symbolName)
```
You could use expression body here instead. Also please apply my previous suggestions here too.

sharwell replied:
📝 Edited this request to reduce work per https://github.com/dotnet/roslyn-analyzers/pull/1111#discussion_r548020530

pakdev replied:
I assume this is outdated

Evangelink commented on src/Text.Analyzers/UnitTests/IdentifiersShouldBeSpelledCorrectlyTests.cs:
```diff
@@ -1,8 +1,531 @@
+                source,
+                dictionary,
+                VerifyCS.Diagnostic(IdentifiersShouldBeSpelledCorrectlyAnalyzer.TypeRule)
+                    .WithLocation(1, 7)
```
Please don't use this overload, either use the markup notation or `WithSpan`.

pakdev replied:
Switched to markupKey; done

Evangelink commented on src/Text.Analyzers/UnitTests/IdentifiersShouldBeSpelledCorrectlyTests.cs:
```diff
@@ -1,8 +1,531 @@
+            await VerifyCSharpAsync(source, dictionary);
+        }
+
+        [Fact(Skip = "Adding additional files to specific projects is not yet supported")]
```
Do we have a ticket to track this?

pakdev replied:
Figured out how to add additional files to projects

Evangelink commented on src/Text.Analyzers/UnitTests/IdentifiersShouldBeSpelledCorrectlyTests.cs:
```diff
@@ -1,8 +1,531 @@
+                    .WithArguments("Clazz", "Clazz"));
+        }
+
+        [Fact(Skip = "Specifying assembly names is not yet supported")]
```
Hum I think we can do it, pretty sure I did it for some test. Let me try to find back the test.

pakdev replied:
Did you find the test?

pakdev replied:
Nevermind; found an example

Evangelink commented on src/Text.Analyzers/UnitTests/IdentifiersShouldBeSpelledCorrectlyTests.cs:
```diff
@@ -1,8 +1,531 @@
+            await VerifyCSharpAsync(
+                source,
+                VerifyCS.Diagnostic(IdentifiersShouldBeSpelledCorrectlyAnalyzer.NamespaceRule)
+                    .WithLocation(1, 17)
```
Please don't use this overload, either use the markup notation or `WithSpan`.

pakdev replied:
Switched to markupKey; done

Evangelink commented on src/Text.Analyzers/UnitTests/IdentifiersShouldBeSpelledCorrectlyTests.cs:
```diff
@@ -1,8 +1,531 @@
+            await VerifyCSharpAsync(
+                source,
+                VerifyCS.Diagnostic(IdentifiersShouldBeSpelledCorrectlyAnalyzer.NamespaceMoreMeaningfulNameRule)
+                    .WithLocation(1, 17)
```
Please don't use this overload, either use the markup notation or `WithSpan`.

pakdev replied:
Switched to markupKey; done

Evangelink commented on src/Text.Analyzers/UnitTests/IdentifiersShouldBeSpelledCorrectlyTests.cs:
```diff
@@ -1,8 +1,531 @@
+
+        private async Task VerifyCSharpAsync(string source, (string Path, string Text)[] additionalTexts, params DiagnosticResult[] expected)
+        {
+
```
```suggestion
```

sharwell replied:
➡️ Marking resolved per https://github.com/dotnet/roslyn-analyzers/pull/1111#discussion_r548020530

Evangelink commented on src/Text.Analyzers/UnitTests/IdentifiersShouldBeSpelledCorrectlyTests.cs:
```diff
@@ -1,8 +1,531 @@
+            await csharpTest.RunAsync();
+        }
+
+        private static (string Path, string Text) CreateXmlDictionary(IEnumerable<string> recognizedWords, IEnumerable<string> unrecognizedWords = null) =>
+            CreateXmlDictionary("CodeAnalysisDictionary.xml", recognizedWords, unrecognizedWords);
```
```suggestion
        private static (string Path, string Text) CreateXmlDictionary(IEnumerable<string> recognizedWords, IEnumerable<string> unrecognizedWords = null)
            => CreateXmlDictionary("CodeAnalysisDictionary.xml", recognizedWords, unrecognizedWords);
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/UnitTests/IdentifiersShouldBeSpelledCorrectlyTests.cs:
```diff
@@ -1,8 +1,531 @@
+            => $"class {typeName} {{ private string {fieldName}; }}";
+
+        private static string CreateTypeWithEvent(string typeName, string eventName)
+        {
+            return $@"using System;
+
+class {typeName} {{ event EventHandler<string> {eventName}; }}";
+        }
```
```suggestion
            => $@"using System;

class {typeName} {{ event EventHandler<string> {eventName}; }}";
```

sharwell replied:
➡️ Resolving per https://github.com/dotnet/roslyn-analyzers/pull/1111#discussion_r548020530

Evangelink commented on src/Utilities.UnitTests/AdditionalFileProviderTests.cs:
```diff
@@ -0,0 +1,124 @@
+
+namespace Analyzers.Utilities.UnitTests
+{
+
```
```suggestion
```

sharwell replied:
I would avoid adding comments for basic code style issues like this. If the style matters as a policy, it will be enforced during the build via analyzers.

(Comments are allowed for cases where analyzers to support the policy do not exist, but if the analyzer exists and the repository chooses not to use it then it's a sign that the style does not constitute a policy that contributors are required to adhere to.)

sharwell replied:
➡️ I'm going to resolve all comments that fall under this bucket to reduce noise in the pull request.

Evangelink commented on src/Utilities/Compiler/AdditionalFileProvider.cs:
```diff
@@ -0,0 +1,50 @@
+        /// </summary>
+        /// <param name="options">Options passed to a <see cref="DiagnosticAnalyzer"/>.</param>
+        /// <returns>An instance of <see cref="AdditionalFileProvider"/>.</returns>
+        public static AdditionalFileProvider FromOptions(AnalyzerOptions options) => new AdditionalFileProvider(options.AdditionalFiles);
```
```suggestion
        public static AdditionalFileProvider FromOptions(AnalyzerOptions options)
            => new AdditionalFileProvider(options.AdditionalFiles);
```

pakdev replied:
Done

Evangelink commented on src/Utilities/Compiler/Extensions/SourceTextExtensions.cs:
```diff
@@ -0,0 +1,47 @@
+        /// <param name="text">Abstraction for an additional file's contents.</param>
+        /// <param name="parser">Function that will parse <paramref name="text"/> into <typeparamref name="T"/>.</param>
+        /// <returns>Output from <paramref name="parser"/>.</returns>
+        public static T Parse<T>(this SourceText text, Func<StreamReader, T> parser)
```
WOuldn't it be better to have this only returning the `StreamReader`?

pakdev replied:
Why would that be better?

Evangelink replied:
Nevermind my comment, I don't know why I missed the usings and thought you will return some content and leave the stream open.

pakdev commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,103 @@
+        /// </Recognized>
+        /// </code>
+        /// </example>
+        public HashSet<string> RecognizedWords { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```
The `RecognizedWords` collection is at least as big as the main dictionary which currently contains 128941 words. The `UnrecognizedWords` collection is probably less than 100, depending on client.

I doubt any benefit from using array over hashset in the case of small datasets warrants assuming the dataset is _always_ small.

Evangelink replied:
Makes sense! I would still go for the `ImmutableHashSet` to provide the safety that the content isn't going to be changed. If you have noticed a perf hit then I would recommend to add a comment to explain the reasoning for the data collection type.

pakdev replied:
Agreed; done

pakdev commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,103 @@
+        }
+
+        /// <summary>
+        /// Creates a new instance of this class with recognized words loaded from the specified DIC <paramref name="streamReader"/>.
```
A file extension/format of a dictionary. Unfortunately, I can't find any specification that I could link here

Evangelink replied:
@mavasani do you happen to have some link?

mavasani replied:
I can only find https://docs.microsoft.com/en-us/visualstudio/code-quality/how-to-customize-the-code-analysis-dictionary?view=vs-2019, but not sure if it the same as the DIC file used here.

pakdev replied:
Yeah, not the same. A *.dic file is just a list of newline-delimited words

pakdev replied:
Added remarks

pakdev commented:
@Evangelink Thanks for taking the time to review my PR, and I appreciate the insightful feedback. Do you mind responding to my questions so I can finish addressing feedback? Thanks!

Evangelink commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,103 @@
+    /// XML or DIC code analysis dictionaries.
+    /// </summary>
+    /// <Remarks>
+    /// <seealso href="https://docs.microsoft.com/en-us/visualstudio/code-quality/how-to-customize-the-code-analysis-dictionary?view=vs-2019"/>
```
```suggestion
    /// <seealso href="https://docs.microsoft.com/visualstudio/code-quality/how-to-customize-the-code-analysis-dictionary"/>
```

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,103 @@
+        /// <param name="unrecognizedWords">Correctly spelled words that the spell checker will now report.</param>
+        private CodeAnalysisDictionary(IEnumerable<string> recognizedWords, IEnumerable<string> unrecognizedWords)
+        {
+            RecognizedWords = new HashSet<string>(recognizedWords, StringComparer.OrdinalIgnoreCase);
```
I think you will have a warning suggesting to use the `new(...)` syntax.

pakdev replied:
With ImmutableHashSet?

pakdev replied:
No warnings with the current code - closing

Evangelink commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,103 @@
+        /// </Recognized>
+        /// </code>
+        /// </example>
+        public HashSet<string> RecognizedWords { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```
Comment still valid IMO.

pakdev replied:
Sorry, I meant please respond to my responses ;) I've addressed this feedback locally and will be pushing shortly.

pakdev replied:
Done

Evangelink commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,103 @@
+        {
+            var document = XDocument.Load(streamReader);
+            // TODO: Include Deprecated words and acronym CasingExceptions as noted here:
+            // https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1704?view=vs-2019#to-add-words-to-a-custom-dictionary
```
```suggestion
            // https://docs.microsoft.com/visualstudio/code-quality/ca1704#to-add-words-to-a-custom-dictionary
```

pakdev replied:
Done

Evangelink reviewed (commented):
Done! You may want to merge latest master to ensure your code complies with some enforcement we have setup lately.

@mavasani just wondering if this should target the `6.0.1xx-preview1` branch as this is a new feature.

pakdev commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,103 @@
+        /// </Unrecognized>
+        /// </code>
+        /// </example>
+        public HashSet<string> UnrecognizedWords { get; } = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
```
Done

pakdev commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,103 @@
+            string word;
+            while ((word = streamReader.ReadLine()) != null)
+            {
+                if (!string.IsNullOrWhiteSpace(word.Trim()))
```
Done

pakdev commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,103 @@
+            {
+                if (!string.IsNullOrWhiteSpace(word.Trim()))
+                {
+                    recognizedWords.Add(word);
```
No; I want the trimmed word here. Fixed

pakdev commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,103 @@
+
+        public static CodeAnalysisDictionary CreateFromDictionaries(IEnumerable<CodeAnalysisDictionary> dictionaries)
+        {
+            var recognizedWords = dictionaries.Select(x => x.RecognizedWords).Aggregate<IEnumerable<string>>((x, y) => x.Union(y));
```
Nope; removed

pakdev commented:
How do I fix the RS2002 build failure? I'm not seeing a codefix for it.

mavasani commented:
> @mavasani just wondering if this should target the 6.0.1xx-preview1 branch as this is a new feature.

It is fine to target master. This analyzer is going into a separate Text.Analyzers package, which does not ship with .NET SDK, so the target branch is irrelevant.

pakdev commented on src/Text.Analyzers/Core/CodeAnalysisDictionary.cs:
```diff
@@ -0,0 +1,103 @@
+        public static CodeAnalysisDictionary CreateFromXml(StreamReader streamReader)
+        {
+            var document = XDocument.Load(streamReader);
+            // TODO: Include Deprecated words and acronym CasingExceptions as noted here:
```
Done

mavasani commented:
@pakdev @Evangelink Do you believe the PR is close to completion? I am fine to merge it as soon as @Evangelink signs off, @pakdev is happy with the current state of the PR and the build is green (I'll push a commit to fix the build error).

pakdev commented:
@mavasani, do you mind responding to @Evangelink's questions here?
- https://github.com/dotnet/roslyn-analyzers/pull/1111#discussion_r555618658
- https://github.com/dotnet/roslyn-analyzers/pull/1111#discussion_r555614547

And then my question:
- https://github.com/dotnet/roslyn-analyzers/pull/1111#discussion_r556277311

Finally, do you already have a commit that resolves [this](https://github.com/dotnet/roslyn-analyzers/pull/1111#discussion_r556584821)? If not, I can do it quickly.

mavasani commented:
> Finally, do you already have a commit that resolves this? If not, I can do it quickly.

I do not right now. I can spend some time later today or early morning to fix it.

pakdev commented:
[This is the last comment](https://github.com/dotnet/roslyn-analyzers/pull/1111#discussion_r568340515) to be resolved, and maybe it already is.

Evangelink reviewed (approved):
I don't have everything fresh in mind but I see all comments are closed and looking through the code I don't see anything obvious so LGTM.

mavasani commented:
Thank you so much @Evangelink!

@pakdev Can you please run `msbuild /t:pack` at the root of the repo and push the auto-updated doc file changes? That should fix the build errors.

pakdev commented:
@mavasani Done - hopefully the build passes and we can finally merge this!

mavasani commented:
@pakdev Linux tests are failing.

pakdev commented:
Hmm, so the one test is failing because I'm checking for an English message and the analyzer is running in a localized content. Is there a way I can ignore checking the diagnostic's message?

mavasani commented:
> Hmm, so the one test is failing because I'm checking for an English message and the analyzer is running in a localized content. Is there a way I can ignore checking the diagnostic's message?

No, you should instead fix your tests to use the resource string/descriptor used by the analyzer.

pakdev commented:
> > Hmm, so the one test is failing because I'm checking for an English message and the analyzer is running in a localized content. Is there a way I can ignore checking the diagnostic's message?
> 
> No, you should instead fix your tests to use the resource string/descriptor used by the analyzer.

I do, but the diagnostic has an argument that's populated with an exception's `Message`. I need to get the localized value of the exception's `Message` somehow.

mavasani commented:
@pakdev Test is still failing on spanish.

mavasani commented:
@pakdev Would you be able to fix the test or skip it with a separate tracking issue?

pakdev commented:
> @pakdev Would you be able to fix the test or skip it with a separate tracking issue?

Test should be fixed now

mavasani commented:
@pakdev and @Evangelink Thanks a ton for digging through this PR and showing tons of patience. Really, really appreciate your dedication. Hopefully, we can release a beta package for this soon on NuGet.org and customers can start dogfooding it.

I am merging the PR now.

FreeAndNil commented:
@mavasani Is there a preview of the new analyzer packe available?

mavasani commented:
@FreeAndNil you can pick up the package from this feed: https://dev.azure.com/dnceng/public/_packaging?_a=package&feed=dotnet6&package=Text.Analyzers&protocolType=NuGet&version=3.3.3-beta1.21268.4&view=versions

FreeAndNil commented:
Thank you very much.
Where can I find the documentation for providing my own dictionary (with brand-names etc.)?

mavasani commented:
@FreeAndNil can you try https://docs.microsoft.com/visualstudio/code-quality/ca1704?view=vs-2019#to-add-words-to-a-custom-dictionary? Also tagging @pakdev in case he has any guidance here.

pakdev commented:
@FreeAndNil to add a custom dictionary
1. Create a file with a name that matches this regex: `(?:dictionary|custom).*?\.(?:xml|dic)$`
2. Include the file as an [additional file](https://github.com/dotnet/roslyn/blob/main/docs/analyzers/Using%20Additional%20Files.md). (note: that documentation describes how to include a certain filetype e.g. `CodeAnalysisDictionary`)

FreeAndNil commented:
@pakdev, @mavasani Thank you very much.
One last question: Is there a possibility to turn of the warnings for single letter names (e.g. e)?
This warning is new (compared to the old analyzers) an contradicts other rules (e.g. for IComparable-Implementations with "x" and "y" as parameter names).

pakdev commented:
> @pakdev, @mavasani Thank you very much.
> One last question: Is there a possibility to turn of the warnings for single letter names (e.g. e)?
> This warning is new (compared to the old analyzers) an contradicts other rules (e.g. for IComparable-Implementations with "x" and "y" as parameter names).

@FreeAndNil Out of the box, no. You should create an issue to request this capability, though. Maybe we could use a different diagnostic ID for "unmeaningful" names.
